posts/scala-for-the-arrogant-part-2.md

```diff
@@ -0,0 +1,427 @@
+MOVES.toStream.map(m => (m, m.doMove(path.endState)))
+```
+This code makes use of one of scala's powerful features: **[for comprehensions](https://docs.scala-lang.org/tour/for-comprehensions.html)**, which provide an alternative way to process streams and handle flatMap, map, and filter operations, as well as collating the results into a collection (or simply preserving the stream for further processing). In the first line we define an Integer Iterator with the `to` keyword creating our **JUG_INDEX_RANGE**. We then use that in the for comprehension to say, in the example of the **Empty** action, "for each integer index over our jugs array, create an Empty move for the specific index."
+
+Because pouring requires two jugs, it is a bit more complicated and we see our first **multi-line for comprehension** using curly braces: "For each integer index over our jugs array defining the jug we are *pouring from*, and for each integer index over our jugs array defining the jugs we are *pouring to*, ignore the cases where the fromJug matches the toJug (since we cannot pour a jug into itself), and yield the remaining Pour moves. What you can see here is that the `<-` operator defines a `map` or iterative operation, while the `if` statement inside of the for comprehension signals a `filter` operation. `yield` most directly corresponds to the collect operation we use in Java. 
+
+Finally, we take these three categories of moves and concatenate them into a single list using `++`. Scala provides a variety of operators for concatenating lists, elements to lists, and other operations that return a new immutable list. We don't need to write any helper methods or use any java libraries to do these types of operations as we had to do prior.
+
+We also use parenthesis to return a [tuple](https://www.tutorialspoint.com/scala/scala_tuples.htm), which allows us to combine a Move and State together as our return type. With Java, we would have to couple those parameters together in a class or List.
+
+One important thing to mention: we are returning a [Stream](https://www.scala-lang.org/api/2.12.3/scala/collection/immutable/Stream.html) instead of a Collection. Streams are lazily loaded, meaning we can delay execution of the doMove method and simply return a data structure that can do that evaluation at a later time or, barring some other circumstance, not at all. We'll see how to advantage of that later on.
+
+At this point we have essentially defined the landscape of our water pouring problem in just a few lines of code. It's vastly simpler than our Java definition and, if it does not seem easy to understand now, with more familiarity regarding Scala's syntax it will be. All that is left is to compose our functions and data types to solve our problem:
+
+```scala
+1. def solve(paths: Stream[Path], history: Set[State]): Stream[Path] = paths match {
+2.     case Stream.Empty => Stream()
+3.     case head #:: tail => {
+4.     val next = createPaths(head)
```

Comment:
Slight indentation problem in this block

Reply:
resolved